src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutInfoNode.cs

```diff
@@ -0,0 +1,305 @@
+            FieldDesc field = methodDescOrFieldDesc as FieldDesc;
+
+            Vertex signature = null;
+
+            if (method != null)
+            {
+                Vertex containingType = GetNativeLayoutInfoSignatureForEEType(factory, method.OwningType);
+                Vertex nameAndSig = _writer.GetMethodNameAndSigSignature(
+                    method.Name,
+                    GetNativeLayoutInfoSignatureForMethodSignature(factory, method.GetTypicalMethodDefinition()));
+                Vertex[] args = null;
+                MethodFlags flags = 0;
+
+                if (method.HasInstantiation)
+                {
+                    flags |= MethodFlags.HasInstantiation;
+                    args = new Vertex[method.Instantiation.Length];
+                    for (int i = 0; i < args.Length; i++)
+                        args[i] = GetNativeLayoutInfoSignatureForEEType(factory, method.Instantiation[i]);
+                }
+
+                signature = _writer.GetMethodSignature((uint)flags, 0, containingType, nameAndSig, args);
+            }
+            else if (field != null)
+            {
+                // TODO
+                throw new NotImplementedException("NYI");
+            }
+
+            Debug.Assert(signature != null);
+
+            signature = _ldTokenInfoSection.Place(signature);
+            return signature;
+        }
+
+        public Vertex GetNativeLayoutInfoSignatureForEEType(NodeFactory factory, TypeDesc type)
+        {
+            IEETypeNode typeSymbol = factory.NecessaryTypeSymbol(type);
+            uint typeIndex = _externalReferences.GetIndex(typeSymbol);
+            return _writer.GetExternalTypeSignature(typeIndex);
+        }
+
+        public Vertex GetNativeLayoutInfoSignatureForMethodSignature(NodeFactory factory, MethodDesc method)
+        {
+            MethodCallingConvention methodCallingConvention = default(MethodCallingConvention);
+
+            if (method.Signature.GenericParameterCount > 0)
+                methodCallingConvention |= MethodCallingConvention.Generic;
+            if (method.Signature.IsStatic)
+                methodCallingConvention |= MethodCallingConvention.Static;
+
+            int parameterCount = method.Signature.Length;
+            Vertex returnType = GetNativeLayoutInfoSignatureForTypeSignature(factory, method.Signature.ReturnType);
+            Vertex[] parameters = new Vertex[parameterCount];
+            for (int i = 0; i < parameterCount; i++)
+            {
+                parameters[i] = GetNativeLayoutInfoSignatureForTypeSignature(factory, method.Signature[i]);
+            }
+
+            return _signaturesSection.Place(_writer.GetMethodSigSignature((uint)methodCallingConvention, (uint)method.Signature.GenericParameterCount, returnType, parameters));
+        }
+
+        public Vertex GetNativeLayoutInfoSignatureForMethod(NodeFactory factory, MethodDesc method, NativeWriter writerToUse = null)
```

Comment:
The `writerToUse` argument looks pretty odd. Do we really need it? It should deserve a comment at least.